Objects/odictobject.c

```diff
@@ -806,30 +807,50 @@ _odict_keys_equal(PyODictObject *a, PyODictObject *b)
 {
     _ODictNode *node_a, *node_b;
 
+    // keep operands' state and size to detect undesired mutations
+    const Py_ssize_t size_a = PyODict_SIZE(a);
+    const Py_ssize_t size_b = PyODict_SIZE(b);
+    if (size_a == -1 || size_b == -1) {
+        return -1;
+    }
+    if (size_a != size_b) {
+        // dictionaries with different number of keys can never be equal
+        return 0;
+    }
+
+    const size_t state_a = a->od_state;
+    const size_t state_b = b->od_state;
+
     node_a = _odict_FIRST(a);
     node_b = _odict_FIRST(b);
-    while (1) {
-        if (node_a == NULL && node_b == NULL)
-            /* success: hit the end of each at the same time */
-            return 1;
-        else if (node_a == NULL || node_b == NULL)
-            /* unequal length */
+    while (node_a != NULL && node_b != NULL) {
+        int res = PyObject_RichCompareBool((PyObject *)_odictnode_KEY(node_a),
+                                           (PyObject *)_odictnode_KEY(node_b),
+                                           Py_EQ);
+        if (res < 0) {
```

Comment:
What if use `res <= 0` here?

Reply:
Errr.... good catch, thanks! Didn't think of that (I'm too used to the `if (... <0)` construction since it's "the path for handling errors").

Reply:
I remember why I did that. You might have `== 0` but this doesn't mean that the state might not have been affected. So the `if` must come *after* the `if` on the state. But the `if` on the state should come after the `if` of the `==` in case something more critical happened.

Reply:
It may be not a problem. The code does not crash, and this is only what is needed in such case.

It does not matter, write what is better for you.